README.md
# Luhn algorithm

[![Dart CI](https://github.com/matejhocevar/luhn_algorithm/actions/workflows/dart-ci.yml/badge.svg)](https://github.com/matejhocevar/luhn_algorithm/actions/workflows/dart-ci.yml)
[![codecov](https://codecov.io/gh/matejhocevar/luhn_algorithm/branch/main/graph/badge.svg?token=J1SUECF4ZA)](https://codecov.io/gh/matejhocevar/luhn_algorithm)
[![Pub](https://img.shields.io/pub/v/luhn_algorithm.svg)](https://pub.dartlang.org/packages/luhn_algorithm)
[![License: MIT][license_badge]][license_link]

> Luhn algorithm is an implementation of the famous Luhn algorithm in Dart. Generate, checksum and validate.

Implementation of this package support:

- ⚙️ Generation of values based on your format. Returns as list or iterable.
- 👌 Computing checksum from provided payload.
- 🧪 Validation of existing value.

## Background

The Luhn algorithm, also known as the modulus 10 or mod 10 algorithm, is a simple checksum formula 
used to validate a variety of identification numbers, such as credit card numbers, IMEI numbers, 
Canadian Social Insurance Numbers. The LUHN formula was created in the late 1960s by a group of mathematicians.
Shortly thereafter, credit card companies adopted it. Because the algorithm is in the public domain,
it can be used by anyone. Most credit cards and many government identification numbers
use the algorithm as a simple method of distinguishing valid numbers from mistyped or otherwise incorrect numbers.

It was designed to protect against accidental errors, not malicious attacks.

The algorithm will detect any single-digit error, as well as almost all transpositions of adjacent digits.
It will not, however, detect transposition of the two-digit sequence 09 to 90 (or vice versa).
It will detect most of the possible twin errors (it will not detect 22 <-> 55, 33 <-> 66 or 44 <-> 77).

Good for:

- ✅ Detecting a character that is misread or badly written
- ✅ Presence or absence of leading zeros at the beginning of the number does not modify the checksum
- ✅ Used by MasterCard, American Express, Visa and others

Not good for:

- ❌ Does not detect certain permutations of digits. E.g. `09` <-> `90` (identical checksum)
- ❌ Failure to detect double errors. E.g. `22` <-> `55`

## Quick start

## Install

```bash
dart pub add luhn_algorithm
```

or add manually add it to pubspec.yaml

```bash
dependencies:
  luhn_algorithm: ^1.0.0
```

## Usage

### Generate
This package contains a function for generating valid Luhn values. You can provide your own formatting, non-digit characters will be ignored.

This is specially useful for testing.

---

#### Generate as iterable

You can finetune the results by specifying the following parameters:

- `format`: specified format, where `x` represents random digit (0-9) and `c` control digit which (!) must be always at the last place in the format.
- `n`: number of returned values

All other character specified in the `format` parameter will be remain there but are going to be ignored in the algorithm compute. 

```dart
import 'package:luhn_algorithm/luhn_algorithm.dart';

var iterator = Luhn.generate(n: 3);
for (String value in iterator) {
  print('Generated value is $value');
}
// Generated value is 91207
// Generated value is 25643
// Generated value is 48322

Luhn.generate(format: 'xxxx-xxxx-xxxx-xxxc', n: 10).forEach(print);
// 8766-6207-9825-0454
// 0748-6872-9179-2506
// 7721-4655-2308-2418
// 5357-3522-2909-5388
// 1132-3019-6824-9848
// 5293-2324-7757-8003
// 9901-8998-8519-1219
// 1899-7700-9488-9896
// 0663-3751-6153-3629
// 2410-2528-3254-5799
```

---

#### Generate as list

This function takes the same parameters and `generate` function.

```dart
import 'package:luhn_algorithm/luhn_algorithm.dart';

Luhn.generateList(format: 'xxxx/xx/xx/xxc', n: 3).forEach(print);
// 3326/95/78/028
// 7705/37/94/556
// 0775/73/88/967

print(Luhn.generateList(format: 'xxc', n: 10));
// [968, 075, 760, 497, 174, 463, 372, 133, 455, 836]
```

---

### Checksum
To get checksum of your value use `checksum` function.

```dart
final var checksum = Luhn.checksum('1234');
print(checksum);
// 4
```

---

### Validate
To validate existing value us static method `Luhn.validate`. Code assumes that the last character of
the string is a control digit.

```dart
print(Luhn.validate('12344'));
// true

print(Luhn.validate('1234-1234-1234-1234'));
// false
```

## Special thanks to

- [Hans Peter Luhn](https://en.wikipedia.org/wiki/Hans_Peter_Luhn)

[license_badge]: https://img.shields.io/badge/license-MIT-blue.svg
[license_link]: https://github.com/matejhocevar/luhn_algorithm/LICENSE


CHANGELOG.md
## 1.0.0+1

### Features

* Initial release
* Add support for `generate`
* Add support for `checksum`
* Add support for `validate`

test/luhn_test.dart
import 'package:luhn_algorithm/luhn_algorithm.dart';
import 'package:test/test.dart';

// ignore_for_file:lines_longer_than_80_chars

void main() {
  group('Luhn generate', () {
    test('should throw an exception if control digit is not in format', () {
      expect(
        () => Luhn.generate(format: ''),
        throwsA(isA<InvalidFormatException>()),
      );
      expect(() => Luhn.generate(format: 'x'),
          throwsA(isA<InvalidFormatException>()));
      expect(() => Luhn.generate(format: 'xxxxxx-xxx'),
          throwsA(isA<InvalidFormatException>()));
    });

    test(
        'should throw an exception if control digit is at the end of the format',
        () {
      expect(() => Luhn.generate(format: 'cx'),
          throwsA(isA<InvalidFormatException>()));
      expect(() => Luhn.generate(format: 'xxcxxxx-xxx'),
          throwsA(isA<InvalidFormatException>()));
      expect(() => Luhn.generate(format: 'cxxcxxxx-xxx'),
          throwsA(isA<InvalidFormatException>()));
    });

    test('should return the correct amount of items', () {
      expect(Luhn.generate(n: 1).length, 1);
      expect(Luhn.generate(n: 10).length, 10);
      expect(Luhn.generate(n: 100).length, 100);
    });

    test('should match the set format', () {
      var format = 'xc';
      var regExp = RegExp(r'^[0-9]{2}$');
      expect(regExp.hasMatch(Luhn.generate(format: format, n: 1).first), true);

      format = 'xxxx-xxxx-xc';
      regExp = RegExp(r'^[0-9]{4}-[0-9]{4}-[0-9]{2}$');
      expect(regExp.hasMatch(Luhn.generate(format: format, n: 10).last), true);

      format = 'zxxxx-zzxxxx#zzzxc';
      regExp = RegExp(r'^z[0-9]{4}-zz[0-9]{4}#zzz[0-9]{2}$');
      expect(regExp.hasMatch(Luhn.generate(format: format).first), true);
    });
  });

  group('Luhn generateList', () {
    test('should throw an exception if control digit is not in format', () {
      expect(() => Luhn.generateList(format: ''),
          throwsA(isA<InvalidFormatException>()));
      expect(() => Luhn.generateList(format: 'x'),
          throwsA(isA<InvalidFormatException>()));
      expect(() => Luhn.generateList(format: 'xxxxxx/xxx'),
          throwsA(isA<InvalidFormatException>()));
    });

    test(
        'should throw an exception if control digit is at the end of the format',
        () {
      expect(() => Luhn.generateList(format: 'cx'),
          throwsA(isA<InvalidFormatException>()));
      expect(() => Luhn.generateList(format: 'xxcxxxx-xxx-'),
          throwsA(isA<InvalidFormatException>()));
      expect(() => Luhn.generateList(format: 'cxxcxxxx-xxx-'),
          throwsA(isA<InvalidFormatException>()));
    });

    test('should return the correct amount of items', () {
      expect(Luhn.generateList(n: 2).length, 2);
      expect(Luhn.generateList(n: 9).length, 9);
      expect(Luhn.generateList(n: 999).length, 999);
    });

    test('should match the set format', () {
      var format = r'x$c';
      var regExp = RegExp(r'^[0-9]\$[0-9]$');
      expect(regExp.hasMatch(Luhn.generateList(format: format).first), true);

      format = '123456xc';
      regExp = RegExp(r'^123456[0-9]{2}$');
      expect(
          regExp.hasMatch(Luhn.generateList(format: format, n: 10).last), true);
    });
  });

  group('Luhn checksum', () {
    test('should return correct control digit', () {
      expect(Luhn.checksum('1'), 8);
      expect(Luhn.checksum('1234567890'), 3);
      expect(Luhn.checksum('938207423894723'), 2);
      expect(Luhn.checksum('1-2-3-4'), 4);
      expect(Luhn.checksum('1#2#3#4'), 4);
      expect(Luhn.checksum('12-23-34-45-56-67-78-89-90'), 2);
    });
  });

  group('Luhn validate', () {
    test('should return false if value is null', () {
      expect(Luhn.validate(null), false);
    });

    test('should return false if value length is 2 chars or less', () {
      expect(Luhn.validate(''), false);
      expect(Luhn.validate('1'), false);
    });

    test('should return false if the value is not correct', () {
      expect(Luhn.validate('11'), false);
      expect(Luhn.validate('12345'), false);
      expect(Luhn.validate('123'), false);
      expect(Luhn.validate('4748192090227444'), false);
    });

    test('should return true if the value is correct', () {
      expect(Luhn.validate('4748192090227442'), true);
      expect(Luhn.validate('4748-1920-9022-7442'), true);
      expect(Luhn.validate('4748#1920#9022#7442'), true);
    });
  });
}


lib/src/luhn_algorithm.dart
import 'dart:math';

import 'package:luhn_algorithm/src/exceptions.dart';

// ignore_for_file:avoid_classes_with_only_static_members
// ignore_for_file:lines_longer_than_80_chars

/// Luhn algorithm implementation
class Luhn {
  /// Creates an iterable for [n] Luhn valid string items.
  ///
  /// ```
  ///   Luhn.generate(format: 'xxxx-xxxxxxx-xxxx', n: 10)
  /// ```
  ///
  /// The format of the generated items can be provided with [format] parameter,
  /// where in format:
  ///
  ///     - x = Random digit
  ///     - c = Control digit.
  ///
  /// Throws and [InvalidFormatException] if control digit is missing
  /// or not present at the end of the format. It will also throw and [InvalidFormatException]
  /// if the control digit is implemented on multiple places.
  ///
  /// Number of items can be provided with [n] parameter.
  static Iterable<String> generate({String format = 'xxxxc', int n = 5}) sync* {
    var items = n;
    if ('c'.allMatches(format).length != 1 ||
        format.indexOf('c') != format.length - 1) {
      throw InvalidFormatException(
        'Control digit (c) should be placed at the end of the format string',
      );
    }

    while (items > 0) {
      final payload = format.split('c')[0];
      final value = payload.replaceAllMapped('x', _replaceWithRandomDigit);
      final controlDigit = checksum(_removeNonDigits(value));
      yield '$value$controlDigit';
      items--;
    }
  }

  /// Generates list of [n] items.
  /// ```
  ///   Luhn.generateList(format: 'xxxx-xxxxxxx-xxxx', n: 10)
  /// ```
  /// Items are Luhn valid strings based on this provided [format].
  static List<String> generateList({String format = 'xxxxc', int n = 1}) {
    return generate(format: format, n: n).toList();
  }

  /// Returns a checksum of provided [value]
  ///
  /// ```
  /// Luhn.checksum('1234') == 4
  /// ```
  static int checksum(String value) {
    final clean = _removeNonDigits(value);
    final sum = _compute(clean);
    return (10 - (sum % 10)) % 10;
  }

  /// Validates whether provided [value] is Luhn valid
  ///
  /// ```
  /// Luhn.validate('12344') == true
  /// ```
  static bool validate(String? value) {
    if (value == null) {
      return false;
    }

    final clean = _removeNonDigits(value);
    if (clean.length < 2) {
      return false;
    }

    return _compute(clean, includesCheckDigit: true) % 10 == 0;
  }

  /// Main Luhn algorithm
  static int _compute(String value, {bool includesCheckDigit = false}) {
    var sum = 0;

    /// If last char is checksum digit we skip first multiplication
    var isDouble = !includesCheckDigit;

    /// We want to double every second digit
    for (var i = value.length - 1; i >= 0; i--) {
      final digit = value.codeUnitAt(i) - 48;

      if (isDouble) {
        final doubled = digit * 2;
        if (doubled > 9) {
          /// Same as we would count digits together
          ///   doubled = 16 -> 1 + 6 = 7
          ///   doubled = 16 -> 16 - 9 = 7
          sum += doubled - 9;
        } else {
          sum += doubled;
        }
      } else {
        sum += digit;
      }

      /// Double the value of every second digit
      isDouble = !isDouble;
    }

    return sum;
  }

  // Helpers

  /// Remove all non-digit chars from the string
  static String _removeNonDigits(String str) {
    final nonDigit = RegExp('[^0-9]');
    return str.replaceAll(nonDigit, '');
  }

  /// Replaces all matches with a random digit
  static String _replaceWithRandomDigit(Match _) =>
      '${Random.secure().nextInt(10)}';
}


example/example.dart
import 'package:luhn_algorithm/luhn_algorithm.dart';

// ignore_for_file: avoid_print

void main() {
  // Function generate

  final iterator = Luhn.generate(n: 3);
  for (final value in iterator) {
    print('Generated value is $value');
  }
  // Generated value is 91207
  // Generated value is 25643
  // Generated value is 48322

  Luhn.generate(format: 'xxxx-xxxx-xxxx-xxxc', n: 10).forEach(print);
  // 8766-6207-9825-0454
  // 0748-6872-9179-2506
  // 7721-4655-2308-2418
  // 5357-3522-2909-5388
  // 1132-3019-6824-9848
  // 5293-2324-7757-8003
  // 9901-8998-8519-1219
  // 1899-7700-9488-9896
  // 0663-3751-6153-3629
  // 2410-2528-3254-5799

  // Function generateList

  Luhn.generateList(format: 'xxxx/xx/xx/xxc', n: 3).forEach(print);
  // 3326/95/78/028
  // 7705/37/94/556
  // 0775/73/88/967

  print(Luhn.generateList(format: 'xxc', n: 10));
  // [968, 075, 760, 497, 174, 463, 372, 133, 455, 836]

  // Function checksum

  final checksum = Luhn.checksum('1234');
  print(checksum);
  // 4

  // Function validate

  print(Luhn.validate('12344'));
  // true

  print(Luhn.validate('1234-1234-1234-1234'));
  // false
}
